### PR TITLE
ci: Refactor frontend.yml CI workflow for clarity and maximum file coverage

### PR DESCRIPTION
**Pull request by ryan953**
This updates the linting tool & file-selection rules to be consistent in dev and in CI. 

Specifically, the main thing I did here is to update `file-filters.yml` so that each tool (eslint, prettier, stylelint) has a specific list of file types to check. If those files, or their configs change, then that trigggers checks at both PR time, and after merge to master. I think the logic is easier to follow now, and it turns out that before it was really easy to skip stylelint for example.

I also updated vscode/settings.json to bring in some stuff from getsentry, so we have linting setup for json which was missing before.

All the files here should basically match their getsentry siblings. We could do better to share things and make sure things stay consistent, so this is a first step. Hopefully we just reduce churn and bugs here altogether.

The getsentry sibling is https://github.com/getsentry/getsentry/pull/13325

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -1,7 +1,6 @@
 **/dist/**/*
 **/vendor/**/*
 **/tests/**/fixtures/**/*
-/scripts/*.js
```
Comment (ryan953) [inline .eslintignore:4]: There are two js files in here, lets just lint them

File: .vscode/settings.json [hunk 2]
```diff
@@ -17,37 +17,7 @@
...
-    "editor.tabSize": 2,
-    "editor.defaultFormatter": "esbenp.prettier-vscode",
-    "editor.codeActionsOnSave": {
-      "quickfix.biome": "explicit",
-      "source.fixAll.eslint": "explicit"
-    }
-  },
-
-  "[javascriptreact]": {
+  "[javascript][typescript][javascriptreact][typescriptreact]": {
```
Comment (ryan953) [inline .vscode/settings.json:20]: weird syntax, but it'll keep things in sync forever which is good.

File: .vscode/settings.json [hunk 3]
```diff
@@ -84,7 +54,7 @@
     }
   },
 
-  "[json]": {
+  "[json][yaml]": {
```
Comment (ryan953) [inline .vscode/settings.json:57]: yaml!

File: package.json [hunk 4]
```diff
@@ -226,11 +226,11 @@
     "lint:js": "eslint static/app tests/js --ext .js,.jsx,.ts,.tsx",
     "lint:css": "stylelint 'static/app/**/*.[jt]sx'",
     "lint:biome": "biome check .",
-    "lint:prettier": "prettier \"**/*.md\" \"**/*.yaml\" \"**/*.{ts,tsx}\" --check",
+    "lint:prettier": "prettier \"**/*.md\" \"**/*.yaml\" \"**/*.[jt]s(x)?\" --check",
```
Comment (ryan953) [inline package.json:229]: need to treat js/ts/jsx/tsx files the same in all the places.

File: .github/workflows/frontend.yml [hunk 5]
```diff
@@ -70,61 +71,47 @@ jobs:
...
-      # Lint entire frontend if:
-      # - this is on main branch
-      # - eslint configuration in repo has changed
-      # - yarn lockfile has changed (i.e. we bump our eslint config)
-      - name: eslint
-        if: steps.eslint.outputs.all-files == 'true'
-        env:
-          # Run relax config on main branch (and stricter config for changed files)
-          SENTRY_ESLINT_RELAXED: 1
+      - name: lint
```
Comment (ryan953) [inline .github/workflows/frontend.yml:74]: This one runs all the lint rules/tools on all the files. 

The tools are currently: biome, prettier, eslint, stylelint.

File: .eslintrc.js [hunk 6]
```diff
@@ -40,7 +33,7 @@ module.exports = {
   ignorePatterns: ['*.json'],
   overrides: [
     {
-      files: ['tests/js/**/*.{ts,js}'],
+      files: ['static/**/*.spec.{ts,js}', 'tests/js/**/*.{ts,js}'],
```
Comment (ryan953) [inline .eslintrc.js:36]: Lots of tests are inside the static folder now

File: .github/file-filters.yml [hunk 7]
```diff
@@ -1,44 +1,69 @@
...
-
-# `frontend_src` filters on files that are frontend changes excluding
-# changes to the tests/ directory.
-# If you want to filter on *all* frontend files, use the `frontend_all` filter.
+sentry_specific_workflow: &sentry_specific_workflow
+  - added|modified: '.github/workflows/frontend.yml'
+sentry_specific_test_files: &sentry_specific_test_files
+  - added|modified: 'tests/js/**/*'
+  - added|deleted|modified: 'fixtures/search-syntax/**/*'
+  - added|deleted|modified: 'fixtures/js-stubs/**/*'
```
Comment (ryan953) [inline .github/file-filters.yml:8]: Everything else in the PR is the same for sentry&getsentry, except this stuff (and where it's referenced down below)

File: .github/file-filters.yml [hunk 8]
```diff
@@ -1,44 +1,69 @@
...
+# Provides list of changed app files to lint against (stylelint)
+lintable_css_in_js_modified: &lintable_css_in_js_modified
+  - added|modified: 'static/**/*.[tj]{s,sx}'
+
+# Trigger for when we must run full lint (stylelint)
+lintable_css_in_js_rules_changed:
+  - added|modified: '.github/file-filters.yml'
+  - added|modified: 'yarn.lock'
+  - added|deleted|modified: 'stylelint.*'
+  - *lintable_css_in_js_modified
```
Comment (ryan953) [inline .github/file-filters.yml:19]: You'll notice that usually there's a pair of file-lists... one to list all the files that we should run the tool against, and another that includes config files which we should watch

File: .github/workflows/frontend.yml [hunk 9]
```diff
@@ -70,61 +71,51 @@ jobs:
...
-          # Run relax config on main branch (and stricter config for changed files)
-          SENTRY_ESLINT_RELAXED: 1
-        run: yarn lint
+      - name: lint
+        if: github.ref == 'refs/heads/master' || needs.files-changed.outputs.lintable_rules_changed != 'true' || needs.files-changed.outputs.lintable_css_in_js_rules_changed != 'true'
+        run: |
+          yarn lint:biome
+          yarn lint:prettier
+          yarn lint:js
+          yarn lint:css
```
Comment (ryan953) [inline .github/workflows/frontend.yml:80]: List this out to be explicit that we're doing the same 4 checks here as below

This is the 'run everything / on master' step
Below are the 'run against changed files' steps

File: .github/workflows/frontend.yml [hunk 10]
```diff
@@ -186,7 +177,7 @@ jobs:
       # it reduces large coverage fluctuations.
       - name: Handle artifacts
         uses: ./.github/actions/artifacts
-        if: ${{ always() && needs.files-changed.outputs.frontend_all == 'true' }}
```
Comment (ryan953) [inline .github/workflows/frontend.yml:189]: The comment above suggests we should do this all the time, and the `frontend_src == 'true'` check is moved up in the tree, not needed here anymore.

File: .vscode/extensions.json [hunk 11]
```diff
@@ -2,14 +2,15 @@
...
-    "ms-python.black-formatter",
     "dbaeumer.vscode-eslint",
+    "esbenp.prettier-vscode",
     "lextudio.restructuredtext",
-    "ziyasal.vscode-open-in-github",
+    "mkhl.direnv",
+    "ms-python.black-formatter",
+    "ms-python.python",
     "timonwong.shellcheck",
-    "tyriar.sort-lines"
```
Comment (ryan953) [inline .vscode/extensions.json:13]: sorted, and added `mkhl.direnv`

This is more important for getsentry. In getsentry we use direnv to find the sentry folder and do some stuff. Added here to for consistency.

File: biome.json [hunk 12]
```diff
@@ -76,7 +76,7 @@
     ]
   },
   "formatter": {
-    "enabled": true,
+    "enabled": false,
```
Comment (ryan953) [inline biome.json:79]: no need to battle with eslint/prettier

**Review by asottile-sentry — CHANGES_REQUESTED** [review]
this changes far too much in one patch to be reviewable

could you split this into multiple, reviewable patches -- in particular please avoid the changes which simply reorder conditions because it's distracting

**Comment by ryan953** [comment]
> this changes far too much in one patch to be reviewable

@asottile-sentry ack, yeah i'll split things up

**Pull request closed without merging**